double wildcard for actor deployment config #19872

svezfaz opened the pull request:
refs #19872

svezfaz committed:
double wildcard for actor deployment config #19872

svezfaz commented:
(Docs changes to follow)

johanandren commented:
OK TO TEST

johanandren commented:
Was it impossible to keep the find method tail recursive, or why the change into a stack based tree traversal?

svezfaz commented:
Essentially this becomes a more generic graph traversal, as a double star can be thought as a branch from a path node to itself. So when a double star is encountered, we can explore two paths and there might be the need for backtracking.
My guess was a tail rec version might not be very readable?

johanandren commented:
Asked mostly because I was curious about the change, would be good to have some idea about the performance implications as well though, compared to the old recursive implementation. AFAIR the mutable Scala collections are not very good in this regard.

svezfaz commented:
Makes sense, let me take a further look and possibly come up with alternatives to compare.
It's a nice little problem to look at :)

ktoso commented on akka-actor-tests/src/test/scala/akka/util/WildcardTreeSpec.scala:
```diff
@@ -0,0 +1,85 @@
+  }
+
+  private implicit def strToIterator(s: String): Iterator[String] = List(s).iterator
+  private implicit def listToIterator[T](l: List[T]): Iterator[T] = l.iterator
```
please no conversions for such simple things

ktoso replied:
Oh well, it's in a test... don't mind as much then.

drewhk replied:
still, if it is not super-ugly, I would refrain from implicits as they might confuse future maintainers of the test.

ktoso commented on akka-actor/src/main/scala/akka/util/WildcardTree.scala:
```diff
@@ -18,15 +18,26 @@ private[akka] final case class WildcardTree[T](data: Option[T] = None, children:
       }
     }
+    WildcardTree()
+  }
```
Would really like to see a benchmark comparing how much we changed perf of the normal case by introducing this. The mutable collection likely has quite some cost. You can write benchmarks in `akka-bench-jmh`, let me know if you have trouble (refer to JMH documentation for how to write one, or just ask here))

patriknw replied:
yeah, it's going to be a challenge to make this performant

is there a way we could cache paths that don't have any entries and thereby only visit them once (that could perhaps be done in the `Deployer` if it's not a concern of the WildcardTree itself)?

drewhk replied:
I am fine if the normal case is not much affected and the generic case is somewhat slower. It would be nice to see how much slower actually.

ktoso commented:
Thanks in advance!

patriknw commented on akka-actor/src/main/scala/akka/util/WildcardTree.scala:
```diff
@@ -18,15 +18,26 @@ private[akka] final case class WildcardTree[T](data: Option[T] = None, children:
-        case None         ⇒ WildcardTree()
+  def find(elems: Iterator[String]): WildcardTree[T] = {
+    case class BFSState(wildcardTree: WildcardTree[T], elems: List[String])
+    val queue = mutable.Queue(BFSState(this, elems.toList))
```
`elems` is already a `List`

patriknw replied:
sorry, looked at wrong `elems`

patriknw replied:
`java.util.LinkedList` might be more performant

drewhk replied:
`ArrayList` even. (no reason for using linked-lists in 99% of the time)

drewhk replied:
sorry, it is used as a Queue, then ArrayDeque?.

patriknw commented on akka-actor/src/main/scala/akka/util/WildcardTree.scala:
```diff
@@ -18,15 +18,26 @@ private[akka] final case class WildcardTree[T](data: Option[T] = None, children:
-        case Some(branch) ⇒ branch.find(elems)
-        case None         ⇒ WildcardTree()
+  def find(elems: Iterator[String]): WildcardTree[T] = {
+    case class BFSState(wildcardTree: WildcardTree[T], elems: List[String])
```
define this class in the companion

patriknw commented on akka-actor/src/main/scala/akka/util/WildcardTree.scala:
```diff
@@ -18,15 +18,26 @@ private[akka] final case class WildcardTree[T](data: Option[T] = None, children:
+
+        kids get state.elems.head foreach explore
+        kids get "*" foreach explore
+        kids get "**" foreach { kid ⇒ explore(kid); exploreYourself() }
```
we must avoid allocations and that means pattern match instead of foreach

patriknw commented on akka-actor/src/main/scala/akka/util/WildcardTree.scala:
```diff
@@ -18,15 +18,26 @@ private[akka] final case class WildcardTree[T](data: Option[T] = None, children:
+    val queue = mutable.Queue(BFSState(this, elems.toList))
+    while (queue.nonEmpty) {
+      val state = queue.dequeue()
+      if (state.elems.isEmpty) return state.wildcardTree
```
no `return`, please (I was staring at the last line `WildcardTree()` and could not understand how this could work ;-))

patriknw commented:
very nice feature @svezfaz, thanks

svezfaz commented:
I've prepared a tailrec version of the BFS function, see below. It uses immutable.Queue. It looks like a better choice, but I would like to try the benchmarking tool on both at some point..

```
  def find(elems: Iterator[String]): WildcardTree[T] = {

    @tailrec
    def findWithBFS(nodes: Queue[(WildcardTree[T], List[String])]): WildcardTree[T] = nodes match {
      case (next, el) +: _ if el.isEmpty ⇒ next
      case (next, el) +: tail ⇒
        val kids = next.children
        val newNodes = tail enqueue
          List(kids.get(el.head), kids.get("*"), kids.get("**"), kids.get("**").map(_ ⇒ next))
          .flatten.map(_ -> el.tail)
        findWithBFS(newNodes)
      case _ ⇒ WildcardTree()
    }

    findWithBFS(Queue(this -> elems.toList))
  }

```

johanandren commented:
A thought that occurred to me, would maybe make sense to only allow the double wildcard at the end of the path? 

I can totally see that use case - we want every child under `/user/some/path/**` to have a specific config, but I find `/some/**/path/` much much harder to reason about. Is there a good use case I'm missing?

svezfaz commented:
Got back on this issue, played around with jmh and written 2 functions: 

1) a `find` that handles `**` in all positions of a path (basically a DFS traversal)

```
  def findWithWildcards(elems: Iterator[String]): WildcardTree[T] = {

    @tailrec
    def findWithDFS(nodes: List[Option[(WildcardTree[T], List[String])]]): WildcardTree[T] = nodes match {
      case Some((next, Nil)) :: _ ⇒ next
      case None :: tail           ⇒ findWithDFS(tail)
      case Some((next @ WildcardTree(_, kids), el)) :: tail ⇒
        findWithDFS(
          (kids.get(el.head) orElse kids.get("*")).map(_ → el.tail) ::
            kids.get("**").map(_ → el.tail) ::
            kids.get("**").map(_ ⇒ next → el.tail) ::
            tail)
      case _ ⇒ WildcardTree()
    }

    findWithDFS(List(Some(this → elems.toList)))
  }

```

2) a `find` that does handle only `**` at the end of a path.

```
  @tailrec def findWithTrailingWildcards(elems: Iterator[String]): WildcardTree[T] =
    if (!elems.hasNext) this
    else children.get("**").filter(_.children.isEmpty) match {
      case Some(branch) ⇒ branch
      case None ⇒
        children.get(elems.next()) orElse children.get("*") match {
          case Some(branch) ⇒ branch.findWithTrailingWildcards(elems)
          case None         ⇒ WildcardTree()
        }
    }
```

On the benchmark I created the results are as follows

```
/*
[info] Benchmark                                        Mode  Cnt   Score   Error  Units
[info] WildcardTreeBenchmark.find                       avgt   30  32.162 ± 2.431  us/op
[info] WildcardTreeBenchmark.findWithWildcards          avgt   30  88.184 ± 2.604  us/op
[info] WildcardTreeBenchmark.findWithTrailingWildcards  avgt   30  41.750 ± 2.657  us/op
 */
```
 
I put everything in https://github.com/svezfaz/akka/commit/ca7ece3577f182abbd3fe948b21be2d2c5a4bed6
The function that handles only trailing wildcards is going to be enough in most situations, however:
- its behaviour will differ from (e.g.) Ant of GitIgnore path matchers
- to avoid backtracking and make it faster, double wildcard matches will have precedence over exact matches (meaning `/a/**` will win over `/a/b`). I'm not sure if this is acceptable.

Any suggestion on which way to choose?

drewhk commented on akka-actor-tests/src/test/scala/akka/actor/DeployerSpec.scala:
```diff
@@ -186,6 +198,18 @@ class DeployerSpec extends AkkaSpec(DeployerSpec.deployerConf) {
       assertRouting("/somewildcardmatch/some", ScatterGatherFirstCompletedPool(nrOfInstances = 1, within = 2 seconds), "/*/some")
     }
 
+    "be able to use double wildcards" in {
```
negative tests? (i.e. tests for non-matches)

svezfaz commented:
There are test failures coming from DeflateSpec, which seems not affected by the PR changes.
Do you know if failures for this test are cropping up anywhere else?
Looks like the timeouts there were increased recently though...

ktoso commented:
Thanks for checking, let's run again and I'll try to get time to review soon (a bit lacking on time due to scaladays this week - sorry for any eventual delays)

ktoso commented:
PLS BUILD

patriknw commented:
That was due to a long GC. It seems like something has been merged to master that results in long GC pauses. We need to investigate that.

```
4071.009: [Full GC (Ergonomics) [PSYoungGen: 293376K->97187K(491520K)] [ParOldGen: 1398156K->1398156K(1398272K)] 1691532K->1495344K(1889792K), [Metaspace: 504323K->504323K(1478656K)], 4.3114312 secs] [Times: user=114.65 sys=0.64, real=4.31 secs] 
```

patriknw commented on akka-actor-tests/src/test/scala/akka/actor/DeployerSpec.scala:
```diff
@@ -203,6 +230,11 @@ class DeployerSpec extends AkkaSpec(DeployerSpec.deployerConf) {
     }
 
+    def assertNoRouting(service: String): Unit = {
+      val deployment = system.asInstanceOf[ActorSystemImpl].provider.deployer.lookup(service.split("/").drop(1))
```
`asInstanceOf[ExtendedActorSystem]` should be enough, I think

patriknw commented on akka-actor/src/main/scala/akka/util/WildcardTree.scala:
```diff
@@ -18,15 +18,24 @@ private[akka] final case class WildcardTree[T](data: Option[T] = None, children:
-        case Some(branch) ⇒ branch.find(elems)
-        case None         ⇒ WildcardTree()
-      }
+  def find(elems: Iterator[String]): WildcardTree[T] = {
```
The way we juggle the elems is amazingly inefficient. `ActorPath.elements` creates a `List` which we convert to an `Iterator` in the `lookup` and then we convert it back to a `List` here.
I think that can be improved by adding a `private[akka] def find(elems: List[String]): WildcardTree[T]` here and use that in `Deployer.lookup(ActorPath)`.

Perhaps we should also consider a `Map` cache for the `lookup` but that can be done in a separate PR.

patriknw replied:
I'm worried about performance and the amount of objects that are created. This is invoked for each and every actor that is created (and we say that actor creation should be cheap). It wasn't good before either.

Perhaps we should introduce the cache in the `Deployer` now, and if so what kind of cache should be used?

svezfaz commented:
Backtracked to terminal double wildcard only. This is the simplest solution and will cover the majority of cases.

Any thoughts?

Docs changes to follow.

patriknw commented:
I think it's good to limit to that.

johanandren commented:
OK TO TEST

johanandren commented:
PLS BUILD

drewhk commented:
I agree with the terminal match.

johanandren commented:
@svezfaz was the original benchmark for the trailing wildcards the exact same impl as now? if not, how does it fare?

svezfaz commented:
@johanandren the implementation was different, I've changed to make double wildcard matches have lower priority than any other match.
I have updated the benchmark branch and re-run, results are very similar though

[info] Benchmark                                        Mode  Cnt    Score    Error  Units
[info] WildcardTreeBenchmark.find                       avgt   30   31.546 ±  3.561  us/op
[info] WildcardTreeBenchmark.findWithTrailingWildcards  avgt   30   40.137 ±  2.123  us/op

johanandren commented:
Ouch that is almost a third slower. I think it would be nice to see if it can get closer to original speed (I misread the original bench as the ops/us kind rather than us/op).

johanandren commented:
MiMa warnings:
```
[info] akka-actor: found 3 potential binary incompatibilities while checking against com.typesafe.akka:akka-actor_2.11:2.3.11  (filtered 216)
[error]  * method lookup(scala.collection.Iterator)scala.Option in class akka.actor.Deployer in current version does not have a correspondent with same parameter signature among (scala.collection.Iterable)scala.Option, (akka.actor.ActorPath)scala.Option
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.actor.Deployer.lookup")
[error]  * method apply()akka.util.WildcardTree in object akka.util.WildcardTree does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.util.WildcardTree.apply")
[error]  * method find(scala.collection.Iterator)akka.util.WildcardTree in class akka.util.WildcardTree does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.util.WildcardTree.find")
```

All three should be safe to filter as both `Deployer` and `WildcardTre` is internal APIs

drewhk commented on project/MiMa.scala:
```diff
@@ -934,7 +935,12 @@ object MiMa extends AutoPlugin {
+        ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.javadsl.RunnableGraph.withAttributes"),
+
+        // #19872 double wildcard for actor deployment config
+        ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.actor.Deployer.lookup"),
```
hrm, is this private API?

johanandren replied:
`private[akka] class Deployer`

drewhk replied:
great!

drewhk commented on project/MiMa.scala:
```diff
@@ -934,7 +935,12 @@ object MiMa extends AutoPlugin {
+
+        // #19872 double wildcard for actor deployment config
+        ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.actor.Deployer.lookup"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.util.WildcardTree.apply"),
```
aren't these public API?

johanandren replied:
`private[akka] final case class WildcardTree`

drewhk replied:
great!

drewhk commented:
LGTM

johanandren commented:
@drewhk what about the performance implication?

drewhk commented:
WDYM? If it is too slow, then this cannot be merged and the PR should be closed.

patriknw commented:
Yeah, don't merge if perf is not ggod enough. Wdyt about the caching that I proposed initially?

svezfaz commented:
The latest version 
1. runs the existing code to look for exact/single wildcard matches
2. If nothing is found, it checks against all double wildcard configs
This should guarantee unchanged performance for all existing projects, where double WC are not used.

@patriknw if we cache in the Deployer, what could the key be? We cannot use ActorPath. If we start deconstructing the path and matching sections of it, we end up with a tree-like cache, which is what we already have. I might be missing something obvious here..

patriknw commented on akka-actor/src/main/scala/akka/util/WildcardTreeIndex.scala:
```diff
@@ -0,0 +1,54 @@
+import scala.annotation.tailrec
+import scala.collection.immutable.HashMap
+
+private[akka] final case class WildcardIndex[T](wildcardTree: WildcardTree[T] = WildcardTree[T](), doubleWildcardTree: WildcardTree[T] = WildcardTree[T]()) {
```
rename the file to match the class name

patriknw commented on akka-actor/src/main/scala/akka/util/WildcardTreeIndex.scala:
```diff
@@ -0,0 +1,54 @@
+      copy(children = children.updated(e, children.getOrElse(e, WildcardTree[T]()).insert(elems, d)))
+    }
+
+  @tailrec def findWithSingleWildcard(elems: Iterator[String]): Option[WildcardTree[T]] =
```
The old implementation used an empty `WildcardTree` to indicate no-match. I think we can use the same here to avoid the `Some` allocations.

patriknw commented on akka-actor/src/main/scala/akka/util/WildcardTreeIndex.scala:
```diff
@@ -0,0 +1,54 @@
+  @tailrec def findWithSingleWildcard(elems: Iterator[String]): Option[WildcardTree[T]] =
+    if (!elems.hasNext) Option(this)
+    else {
+      children.get(elems.next()) orElse children.get("*") match {
```
the `orElse` can be rewritten as a `match` to avoid "thunk" allocation

patriknw commented on akka-actor/src/main/scala/akka/util/WildcardTreeIndex.scala:
```diff
@@ -0,0 +1,54 @@
+
+  def find(elems: Iterable[String]): Option[T] =
+    wildcardTree.findWithSingleWildcard(elems.iterator)
+      .orElse(doubleWildcardTree.findWithTerminalDoubleWildcard(elems.iterator))
```
Make a special case for empty WildcardTree to avoid the `elems.iterator` and traversal thereof. I think it's worth doing that for both `wildcardTree` and `doubleWildcardTree`

patriknw commented:
@svezfaz I have reviewed again. A few more comments to optimize it further. I think after that this should be good enough for performance and can be merged. Thanks

svezfaz commented:
Thanks a lot @patriknw , I have pushed the amendments.

drewhk commented:
Looks reasonable. What is the performance now?

svezfaz commented:
Just run a benchmark, they now look in line with the current implementation.

```
[info] Benchmark                                Mode  Cnt   Score   Error  Units
[info] WildcardIndexBenchmark.find_deepTree     avgt   30  24.136 ± 0.363  us/op
[info] WildcardIndexBenchmark.find_double       avgt   30   1.082 ± 0.010  us/op
[info] WildcardIndexBenchmark.find_single       avgt   30   1.000 ± 0.005  us/op
[info] WildcardIndexBenchmark.find_wideTree     avgt   30   0.114 ± 0.002  us/op
[info] WildcardIndexBenchmark.oldFind_deepTree  avgt   30  24.387 ± 0.383  us/op
[info] WildcardIndexBenchmark.oldFind_double    avgt   30   0.963 ± 0.012  us/op
[info] WildcardIndexBenchmark.oldFind_single    avgt   30   0.978 ± 0.007  us/op
[info] WildcardIndexBenchmark.oldFind_wideTree  avgt   30   0.106 ± 0.001  us/op
[info] 	akka.actor.WildcardIndexBenchmark.find_deepTree_	24	us/op
```

patriknw commented on akka-actor/src/main/scala/akka/util/WildcardIndex.scala:
```diff
@@ -0,0 +1,74 @@
+
+  def find(elems: Iterable[String]): Option[T] = {
+    (forNonEmpty(wildcardTree)(_.findWithSingleWildcard(elems.iterator)) match {
+      case `empty` ⇒ forNonEmpty(doubleWildcardTree)(_.findWithTerminalDoubleWildcard(elems.iterator))
```
that case (and the one in `forNonEmpty`)  will be execute `equals` on the `WildcardTree`, which might be too costly.
Introduce an `isEmpty` method on the `WildcardTree` and use `if` statements.

I could also nitpick on that using `forNonEmpty` allocates the `f` thunk. Optimal performance would be to write the hole thing inline with if statements.

Yeah, performance is more important than readability here.

patriknw commented:
thanks for the update and running the test, looking good, just one more nitpick ;-)

svezfaz commented:
Ran the benchmark again, results are still in line.

johanandren commented:
LGTM!

patriknw commented on akka-actor/src/main/scala/akka/util/WildcardIndex.scala:
```diff
@@ -0,0 +1,81 @@
+      } else {
+        withSingleWildcard
+      }
+    }).data
```
thanks

patriknw commented:
LGTM

patriknw commented:
@svezfaz Thanks for an excellent contribution, and for your patience.
